dsl/flatten.cc

```diff
@@ -0,0 +1,340 @@
+            ENFORCE(method.expr != nullptr);
+            insSeq->stats.emplace_back(std::move(method.expr));
+        }
+        return tree;
+    }
+
+private:
+    ast::ClassDef::RHS_store addMethods(core::Context ctx, ast::ClassDef::RHS_store rhs, core::Loc loc) {
+        if (curMethodSet().methods.size() == 1 && rhs.size() == 1 && ast::isa_tree<ast::EmptyTree>(rhs[0].get())) {
+            // It was only 1 method to begin with, put it back
+            rhs.pop_back();
+            rhs.emplace_back(std::move(popCurMethodDefs()[0].expr));
+            return rhs;
+        }
+
+        auto exprs = popCurMethodDefs();
+        // TODO: remove all this
+        //
+        // this add the nested methods at the appropriate 'staticness level'
+
+        int highestLevel = 0;
+        for (int i = 0; i < exprs.size(); i++) {
+            auto &expr = exprs[i];
+            if (highestLevel < expr.staticLevel) {
+                highestLevel = expr.staticLevel;
+            }
+            // we need to make sure that we keep sends with their attached methods, so fix that up here
+            if (i > 0) {
+                auto send = ast::cast_tree<ast::Send>(exprs[i - 1].expr.get());
+                if (send != nullptr && send->fun == core::Names::sig()) {
+                    exprs[i - 1].staticLevel = expr.staticLevel;
+                }
+            }
+        }
+
+        // these will store the bodies of the `class << self` blocks we create at the end
+        vector<ast::ClassDef::RHS_store> nestedBlocks;
+        for (int level = 2; level <= highestLevel; level++) {
+            nestedBlocks.emplace_back();
+        }
+
+        // this vector contains all the possible RHS target locations that we might move to
+        vector<ast::ClassDef::RHS_store *> targets;
+        // 0 and 1 both go into the class itself
+        targets.emplace_back(&rhs);
+        targets.emplace_back(&rhs);
+        // 2 and up go into the to-be-created `class << self` blocks
+        for (auto &tgt : nestedBlocks) {
+            targets.emplace_back(&tgt);
+        }
+
+        // move everything to its appropriate target
+        for (auto &expr : exprs) {
+            if (auto methodDef = ast::cast_tree<ast::MethodDef>(expr.expr.get())) {
+                methodDef->setIsSelf(expr.staticLevel > 0);
+            }
+            targets[expr.staticLevel]->emplace_back(std::move(expr.expr));
+        }
+
+        // generate the nested `class << self` blocks as needed and add them to the class
+        for (auto &body : nestedBlocks) {
+            auto classDef =
+                ast::MK::Class(loc, loc,
+                               make_unique<ast::UnresolvedIdent>(core::Loc::none(), ast::UnresolvedIdent::Class,
+                                                                 core::Names::singleton()),
+                               {}, std::move(body), ast::ClassDefKind::Class);
+            rhs.emplace_back(std::move(classDef));
+        }
```

Comment:
This whole section is here to be compatible with the current incorrect semantics. Once we break compatibility with the current semantics, we'll be able to delete all this code and simply add everything to the `rhs` (instead of having to partition it out into possibly-to-be-created `class << self` blocks.)